### PR TITLE
add simd horizantal reduce to improve log_softmax and softmax performance on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76279
* #76278
* __->__ #73953

Differential Revision: [D35782875](https://our.internmc.facebook.com/intern/diff/D35782875)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35782875).

**Comment by mingfeima** [comment]
TODO:
- [ ] move this one out and continue to improve bf16 on softmax/log_softmax when last dim

**Comment by mingfeima** [comment]
The target of this stack is to improve some hotspot operators in transformer, e.g. `softmax` and `log_softmax` when dim = -1.

This PR is vectorizing the `vec_reduce_all`, the original impl is a slow path with scalar logic.
Usually the size of last dimension of `softmax` used in transformer is not very big, so if `vec_reduce_all` is not efficient it's going to take a large ratio.

Benchmark data on Xeon 6248, 20 cores per socket, dual socket, 2.50GHz:
### cpu performance on 4 cores run
```
### before
<class 'torch.nn.modules.activation.Softmax'> torch.Size([16, 384, 384]) torch.float32 time: 1.472 ms
<class 'torch.nn.modules.activation.Softmax'> torch.Size([1, 12, 512, 512]) torch.float32 time: 1.725 ms

### after
<class 'torch.nn.modules.activation.Softmax'> torch.Size([16, 384, 384]) torch.float32 time: 1.093 ms
<class 'torch.nn.modules.activation.Softmax'> torch.Size([1, 12, 512, 512]) torch.float32 time: 1.550 ms
```

### cpu performance on 1 core run
```
### before
<class 'torch.nn.modules.activation.Softmax'> torch.Size([16, 384, 384]) torch.float32 time: 2.990 ms
<class 'torch.nn.modules.activation.Softmax'> torch.Size([1, 12, 512, 512]) torch.float32 time: 3.799 ms

### after
<class 'torch.nn.modules.activation.Softmax'> torch.Size([16, 384, 384]) torch.float32 time: 2.684 ms
<class 'torch.nn.modules.activation.Softmax'> torch.Size([1, 12, 512, 512]) torch.float32 time: 3.570 ms
```

**Comment by mingfeima** [comment]
Update: disable mobile build for `VecReduceAllSIMD`, mobile platform will fallback to slow path.

**Comment by mingfeima** [comment]
we don't need to add new test for this one, already covered from
`vec_test_all_types_AVX2`, `vec_test_all_types_AVX512` and `vec_test_all_types_DEFAULT`.

The functional test will iterate array length from 1 to `N = vec::size() * 2 + 2`
* when length < vec::size(); original slow path is used (one by one scalar reduction)
* when length >= vec::size(), path vectorized path is used.

**Comment by frank-wei** [comment]
@frank-wei has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35782875).

**Review by frank-wei — APPROVED** [review]

File: aten/src/ATen/cpu/vec/functional_base.h [hunk 1]
```diff
@@ -27,6 +27,62 @@ inline scalar_t vec_reduce_all(
...
+template <typename scalar_t, typename Op>
+inline scalar_t vec_reduce_all(
+    const Op& vec_fun,
+    vec::Vectorized<scalar_t> acc_vec) {
+  using Vec = vec::Vectorized<scalar_t>;
+  return vec_reduce_all(vec_fun, acc_vec, Vec::size());
+}
+
+// SIMD version on horizontal reduce
+#if defined(CPU_CAPABILITY_AVX512) && !defined(_MSC_VER)
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/cpu/vec/functional_base.h]: ATen's vectorized struct `at::vec::Vectorized<>`  will be compiled 3 times: DEFAULT (the scalar version), AVX2(-mavx2 build) and AVX512 (-mavx512 build). The latter two are the real vectorized instructions and would eventually go to intrisincs. 

`_MSC_VER` refers to MSVC (windows c/c++ compiler), at the moment, `at::vec::Vectorized<>` only have the DEFAULT version on windows. Both AVX2 and AVX512 build are disabled on windows, as shown here:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec/vec256/vec256_float.h#L18

GCC and Clang are OK, otherwise CIs would have failed.

Also this PR is only enabled with GCC version > 5:
```
#if defined(__GNUC__) && (__GNUC__ > 5) && !defined(_MSC_VER) && !defined(C10_MOBILE)
```
(gcc5 did not recognize the intrinsics `_mm256_cvtss_f32`, which causes CI failed previously)
